lib/gmp/gmp.nit

```diff
@@ -0,0 +1,370 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 201 Alexis Laferrière <alexis.laf@xymus.net>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Multi precision integer and rational number using gmp lib
+module gmp
+
+private import native_gmp
+
+redef class Numeric
+
+    fun to_l: Long do return self.to_i.to_l
+
+    fun to_r: Ratio do return self.to_f.to_r
+
+end
+
+redef class Text
+
+    #     assert("10".to_l == 10.to_l)
+    fun to_l: Long do
+        var tmp = new NativeMPZ
+        tmp.set_str(self.to_cstring, 10i32)
+        return new Long(tmp)
+    end
+
+    fun to_r: Ratio do
+        var tmp = new NativeMPQ
+        tmp.set_str self.to_cstring
+        return new Ratio(tmp)
+    end
+end
+
+redef class Float
+    redef fun to_l do
+        var tmp = new NativeMPZ
+        tmp.set_d self
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_d self
+        return new Ratio(tmp)
+    end
+end
+
+redef class Int
+    redef fun to_l do
+        var tmp = new NativeMPZ
+        tmp.set_si self
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_si(self, 1)
+        return new Ratio(tmp)
+    end
+end
+
+class Long
+    super Discrete
+    super Numeric
+    super FinalizableOnce
+
+    redef type OTHER: Long
+
+    private var val: NativeMPZ
+
+    redef fun successor(l) do return self + l.to_l
+    redef fun predecessor(l) do return self - l.to_l
+
+    redef fun hash do return self.to_i
+
+    redef fun <=>(l) do
+        if (val.cmp(l.val)) < 0 then
```

Comment:
You should cache this locally, else it might (will?) be performed twice if `self >= l`